server/src/main/java/org/opensearch/cluster/service/Throttler.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.service;
+
+import org.opensearch.common.AdjustableSemaphore;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Base class for Throttling logic.
+ * It provides throttling functionality over multiple keys.
+ *
+ * @param <T> the type of key on which we want to do throttling.
+ */
+public class Throttler<T> {
```

Comment:
I just wanted to keep it generic so in future if we need throttling functionality somewhere else we can leverage it. 

For master task throttling case, key will be mostly string, dont want to make this base throttler class to stick with string key. In upcoming PR we will have MasterTaskThrottler which will extend this and have string as key. 

Similarly in future, if some other functionality wants to use Throttler class for other type of key they can also use it.

Reply:
I believe we could drop `T` (in favor of String) for now and introduce the change when needed (btw none of the usages in `ThrottlerTests` acknowledges that `Throttler` is actually generic type).

Reply:
IMHO we should keep Throttler generic so we can use it whenever required(without any change). I am happy to change it to `String` but want to understand the rational behind keeping it `String` instead of `T`. 

(I checked in `ThrottlerTests`, by default it was treating it as `Object` type as key. Corrected and added `String` type in key)

Reply:
> IMHO we should keep Throttler generic so we can use it whenever required(without any change). 

There is no use case for it, you just confirmed that in previous comment (https://github.com/opensearch-project/OpenSearch/pull/3527#discussion_r915815572): `I just wanted to keep it generic so in future if we need throttling functionality somewhere else we can leverage it.`. The change could be done easily **when the need comes**, possibly never.

Reply:
Thanks for quick reply. 

Will remove the generics and kept it as String.